src/emmy/complex/impl.cljc

```diff
@@ -0,0 +1,616 @@
+;;  **/
+
+(ns emmy.complex.impl
+  "This namespace provides the primitive implementation of complex
+   arithmetic for Emmy. The authors are indebted to Robert Eisele's
+   `Complex.js` implementation from which we have borrowed freely and
+   which is licensed as follows:
+
+   * Copyright (c) 2020, Robert Eisele (robert@xarg.org)
+   * Dual licensed under the MIT or GPL Version 2 licenses."
+  (:refer-clojure :exclude [abs zero?])
+  (:require [emmy.generic :as g]
+            [emmy.util :as u]
+            [emmy.value :as v]))
+
+(declare equal?)
+
+(deftype Complex [re im]
+  v/IKind
+  (kind [_] :emmy.complex/complex)
+
+  v/Numerical
+  (numerical? [_] true)
+
+  #?@(:clj [Object
+            (equals [a b] (equal? a b))]
+      :cljs [IEquiv
+             (-equiv [a b] (equal? a b))
+
+             IPrintWithWriter
+             (-pr-writer
+              [z writer _]
+              (write-all
+               writer
+               "#emmy/complex "
+               (str [(.-re z) (.-im z)])))]))
+
+(def ZERO (Complex. 0 0))
+(def ONE (Complex. 1 0))
+(def I (Complex. 0 1))
+(def -I (Complex. 0 -1))
+(def INFINITY (Complex. ##Inf ##Inf))
+(def NAN (Complex. ##NaN ##NaN))
+(def LN2 (Math/log 2))
+
+(def ^:private PI:2 (/ Math/PI 2))
+(def ^:private PI:4 (/ Math/PI 4))
+
+(defn equal?
+  "Test for complex equality with another object."
+  [^Complex a b]
+  (cond (instance? Complex b)
+        (and (v/= (.-re a) (.-re b))
+             (v/= (.-im a) (.-im b)))
+
+        (v/real? b)
+        (and (g/zero? (.-im a))
+             (v/= (.-re a) b))
+
+        :else (v/= a b)))
+
+(defn zero?
+  "Determines whether or not a complex number is at the zero pole of the
+  Riemann sphere."
+  [^Complex z]
+  (and (g/zero? (.-re z))
+       (g/zero? (.-im z))))
+
+(defn nan?
+  "Determines whether a complex number is not on the Riemann sphere."
+  [^Complex z]
+  (or (u/nan? (.-re z))
+      (u/nan? (.-im z))))
+
+(def ^:no-doc cos-1-square-terms
+  "For the origin of these constants, see the related material
+   in [[emmy.series-test/cos-1-square-terms]]."
+  [4.779477332387385E-14 -1.147074559772972E-11 2.08767569878681E-9 -2.755731922398589E-7
+   2.48015873015873E-5 -0.001388888888888889 0.04166666666666667 -0.5 0.0])
+
+(defn cos-1
+  "Computes $\\cos(x)-1$ using Taylor series if $|x|\\le{\\pi\\over 4}$),
+   otherwise just subtracts one from `(Math/cos x)`. Doing the latter
+   when x is small squanders significant digits."
+  [x]
+  (cond (g/zero? x) (g/zero-like x)
+        (> (g/abs x) PI:4) (- (Math/cos x) 1)
+        :else
+        (let [xx (* x x)]
+          (reduce (fn [a b] (+ (* a xx) b)) cos-1-square-terms))))
+
+(defn real
+  [^Complex z]
+  (.-re z))
+
+(defn imaginary
+  [^Complex z]
+  (.-im z))
+
+(defn abs
+  "Calculate the magnitude of the complex number"
+  [^Complex z]
+  (let [x (.-re z)
+        y (.-im z)
+        a (g/abs x)
+        b (g/abs y)]
+    (cond (and (< a 3000) (< b 3000))
```

Comment:
can we document this? presumably this is to kill overflow but I'd love a note about the choice, if you understand why.

Reply:
I have replaced the implementation with one cribbed from the GNU ISO C++ stdlib. That implementation pre-divides by the greater of the components and then corrects the result. It's much clearer what's going on.